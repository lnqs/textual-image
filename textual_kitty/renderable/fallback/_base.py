"""Provides functionality to render images with Rich as unicode characters."""

import logging
from pathlib import Path
from typing import override

from PIL import Image as PILImage

from textual_kitty.geometry import Size
from textual_kitty.renderable._base import _Base
from textual_kitty.terminal import TerminalSizeInformation

logger = logging.getLogger(__name__)


class _FallbackBase(_Base):
    """Rich renderable to render images in terminal.

    This Renderable renders an image as unicode characters. Used as fallback if other protocols are not available.
    """

    @override
    def __init__(self, image: str | Path | PILImage.Image, width: int | None = None) -> None:
        super().__init__(image, width)
        self._prepared_image: PILImage.Image | None = None

    @override
    def is_prepared(self, size: Size, term_size: TerminalSizeInformation) -> bool:
        if self._prepared_image and self._prepared_image.size == self._calculate_image_size_for_render_size(
            size, term_size
        ):
            return True
        return False

    @override
    def prepare(self, size: Size, term_size: TerminalSizeInformation) -> None:
        logger.debug("Scaling fallback image for rendering")
        self._prepared_image = self._prepare_image(size, term_size)

    @override
    async def prepare_async(self, size: Size, term_size: TerminalSizeInformation) -> None:
        logger.debug("Scaling fallback image for rendering asynchrounouly")
        self._prepared_image = await self._run_in_thread(lambda: self._prepare_image(size, term_size))

    def _prepare_image(self, size: Size, term_size: TerminalSizeInformation) -> PILImage.Image:
        raise NotImplementedError()

    def _calculate_image_size_for_render_size(self, render_size: Size, term_size: TerminalSizeInformation) -> Size:
        raise NotImplementedError()
