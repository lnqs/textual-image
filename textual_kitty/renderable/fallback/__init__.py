"""Rich `Renderable`s to display images as unicode characters.

To be used if no other protocols are available
"""
