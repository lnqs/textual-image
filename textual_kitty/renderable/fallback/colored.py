"""Provides functionality to render images with Rich as colored unicode characters."""

from contextlib import nullcontext
from typing import Tuple, cast, override

from PIL import Image as PILImage
from rich.color import Color
from rich.color_triplet import ColorTriplet
from rich.console import RenderResult
from rich.segment import Segment
from rich.style import Style

from textual_kitty.geometry import Size
from textual_kitty.renderable.fallback._base import _FallbackBase
from textual_kitty.terminal import TerminalSizeInformation


class Image(_FallbackBase):
    """Rich renderable to render images in terminal.

    This Renderable renders an image as colored unicode characters. Used as fallback if other protocols are not available.
    """

    @override
    def _prepare_image(self, size: Size, term_size: TerminalSizeInformation) -> PILImage.Image:
        with nullcontext(self.image) if isinstance(self.image, PILImage.Image) else PILImage.open(self.image) as image:  # type: ignore
            return cast(PILImage.Image, image.resize(self._calculate_image_size_for_render_size(size, term_size)))

    @override
    def _calculate_image_size_for_render_size(self, render_size: Size, term_size: TerminalSizeInformation) -> Size:
        return Size(int(render_size.width), int(render_size.height / (term_size.cell_width / term_size.cell_height)))

    @override
    def _render(self) -> RenderResult:
        if not self._prepared_image:
            return

        for y in range(0, self._prepared_image.height, 2):
            for x in range(self._prepared_image.width):
                upper = self._get_pixel(x, y)
                lower = self._get_pixel(x, y + 1) if y + 1 < self._prepared_image.height else None
                yield Segment("â–€", style=Style(color=upper, bgcolor=lower))
            yield Segment("\n")

    def _get_pixel(self, x: int, y: int) -> Color:
        if not self._prepared_image:
            return Color.default()

        data = self._prepared_image.getpixel((x, y))
        return Color.from_triplet(ColorTriplet(*cast(Tuple[int, int, int], data)))
