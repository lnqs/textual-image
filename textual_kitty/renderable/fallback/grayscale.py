"""Provides functionality to render images with Rich as unicode characters."""

from contextlib import nullcontext
from typing import cast, override

from PIL import Image as PILImage
from rich.console import RenderResult
from rich.segment import Segment

from textual_kitty.geometry import Size
from textual_kitty.renderable.fallback._base import _FallbackBase
from textual_kitty.terminal import TerminalSizeInformation

CHARACTERS = [
    "█",  # FULL BLOCK
    "▓",  # DARK SHADE
    "▒",  # MEDIUM SHADE
    "░",  # LIGHT SHADE
    " ",  # SPACE
]

CHARACTER_LOOKUP = {int(255 / len(CHARACTERS) * i): c for i, c in enumerate(CHARACTERS)}


class Image(_FallbackBase):
    """Rich renderable to render images in terminal.

    This Renderable renders an image as unicode characters. Used as fallback if other protocols are not available.
    """

    @override
    def _prepare_image(self, size: Size, term_size: TerminalSizeInformation) -> PILImage.Image:
        with nullcontext(self.image) if isinstance(self.image, PILImage.Image) else PILImage.open(self.image) as image:  # type: ignore
            image = image.resize(self._calculate_image_size_for_render_size(size, term_size))
            image = image.convert("L")
            return cast(PILImage.Image, image)

    @override
    def _calculate_image_size_for_render_size(self, render_size: Size, term_size: TerminalSizeInformation) -> Size:
        return Size(
            int(render_size.width), int(render_size.height / 2 / (term_size.cell_width / term_size.cell_height))
        )

    @override
    def _render(self) -> RenderResult:
        if not self._prepared_image:
            return

        for y in range(0, self._prepared_image.height):
            for x in range(self._prepared_image.width):
                character = self._get_pixel(x, y)
                yield Segment(character)
            yield Segment("\n")

    def _get_pixel(self, x: int, y: int) -> str:
        if not self._prepared_image:
            return " "

        pixel = cast(int, self._prepared_image.getpixel((x, y)))
        brightness = 255 - pixel
        index = int(255 / len(CHARACTER_LOOKUP)) * int(brightness / int(255 / len(CHARACTER_LOOKUP)))
        return CHARACTER_LOOKUP[index]
