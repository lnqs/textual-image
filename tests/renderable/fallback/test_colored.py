from typing import cast

from PIL import Image as PILImage
from rich.color import Color
from syrupy.assertion import SnapshotAssertion

from tests.data import SIZE_TEN, TERM_SIZE, TEST_IMAGE
from tests.util import render
from textual_kitty.renderable.fallback.colored import Image


def test_prepare() -> None:
    renderable = Image(TEST_IMAGE)
    renderable.prepare(SIZE_TEN, TERM_SIZE)
    assert cast(PILImage.Image, renderable._prepared_image).size == (10, 22)


async def test_prepare_async() -> None:
    renderable = Image(TEST_IMAGE)
    await renderable.prepare_async(SIZE_TEN, TERM_SIZE)
    assert cast(PILImage.Image, renderable._prepared_image).size == (10, 22)


def test_calculate_image_size_for_render_size() -> None:
    renderable = Image(TEST_IMAGE)
    assert renderable._calculate_image_size_for_render_size(SIZE_TEN, TERM_SIZE) == (10, 22)


def test_render_not_prepared() -> None:
    renderable = Image(TEST_IMAGE)
    assert list(renderable._render()) == []


def test_render(snapshot: SnapshotAssertion) -> None:
    renderable = Image(TEST_IMAGE, width=4)
    assert render(renderable) == snapshot


def test_get_pixel_failure() -> None:
    renderable = Image(TEST_IMAGE)
    assert renderable._get_pixel(0, 0) == Color.default()
