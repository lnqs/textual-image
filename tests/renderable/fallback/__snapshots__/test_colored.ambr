# serializer version: 1
# name: test_render
  '''
  [38;2;52;50;86;48;2;58;46;70mâ–€[0m[38;2;89;74;77;48;2;166;102;74mâ–€[0m[38;2;102;103;122;48;2;175;122;108mâ–€[0m[38;2;70;109;180;48;2;104;141;201mâ–€[0m
  [38;2;62;37;40;48;2;43;36;36mâ–€[0m[38;2;114;77;62;48;2;34;35;41mâ–€[0m[38;2;108;85;84;48;2;45;38;46mâ–€[0m[38;2;62;78;114;48;2;32;18;25mâ–€[0m
  
  '''
# ---
